prog/mutation.go

```diff
@@ -419,49 +467,124 @@ type mutationArgs struct {
-			return // For structs only individual fields are updated.
-		}
-		// These special structs are mutated as a whole.
-		ctx.Stop = true
-	case *UnionType:
-		if ma.target.SpecialTypes[typ.Name()] == nil && len(typ.Fields) == 1 || ignoreSpecial {
-			return
-		}
-		ctx.Stop = true
-	case *ArrayType:
-		// Don't mutate fixed-size arrays.
-		if typ.Kind == ArrayRangeLen && typ.RangeBegin == typ.RangeEnd {
 			return
 		}
-	case *CsumType:
-		return // Checksum is updated when the checksummed data changes.
-	case *ConstType:
-		return // Well, this is const.
-	case *BufferType:
-		if typ.Kind == BufferString && len(typ.Values) == 1 {
-			return // string const
-		}
 	case *PtrType:
 		if arg.(*PointerArg).IsSpecial() {
 			// TODO: we ought to mutate this, but we don't have code for this yet.
 			return
 		}
 	}
+
+	prio, ctx.Stop, canMutate = arg.Type().getMutationPrio()
+
 	typ := arg.Type()
-	if typ == nil || typ.Dir() == DirOut || !typ.Varlen() && typ.Size() == 0 {
+	if typ == nil || typ.Dir() == DirOut || !typ.Varlen() && typ.Size() == 0 || !canMutate {
 		return
 	}
 	ma.args = append(ma.args, arg)
 	ma.ctxes = append(ma.ctxes, *ctx)
+	ma.priorities = append(ma.priorities, prio)
+}
+
+// Assign a priority for each type. The boolean is the reference type and it has
+// the minimum priority, since it has only two possible values.
+func (t *IntType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if (t.RangeEnd - t.RangeBegin) == 1 {
+		return minPriority, true, true
+	}
+
+	bitsCount := math.Log2((float64(t.TypeSize * 8)))
+	if t.RangeBegin == t.RangeEnd {
+		return bitsCount + 2, true, true
+	}
+
+	// an integer with a range could have a higher impact
+	return bitsCount + 4, true, true
+}
+
+func (t *StructType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return maxPriority, true, true
+}
+
+func (t *UnionType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if len(t.Fields) == 1 {
+		return minPriority, false, false
+	}
+
+	return maxPriority, true, true
+}
+
+func (t *FlagsType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if len(t.Vals) == 0 { // No value to mutate
+		return minPriority, true, false
+	}
+
+	// when bitmask is true we should a have a higher priority because we can have bitwise operations
+	if t.BitMask {
+		bitsCount := math.Log2((float64(t.TypeSize * 8)))
+		return bitsCount + 3, true, true // a higher priority than an integer without range
+	}
+
+	return 0.6 * maxPriority, true, true
+}
+
+func (t *PtrType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 3, false, true
+}
+
+func (t *ConstType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return minPriority, true, false
+}
+
+func (t *CsumType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return minPriority, true, false
+}
+
+// TODO: a better value for the hardcoded priorities
```

Comment:
Move this TODO before all of the `getMutationPrio()` definitions and reword as something like "find a way to estimate optimal priority values".